src/arch/riscv/isa/templates/vector_arith.isa

```diff
@@ -358,6 +358,71 @@ Fault
         return std::make_shared<IllegalInstFault>("VILL is set", machInst);
 
     const int64_t vlmul = vtype_vlmul(machInst.vtype8);
+    const uint32_t num_microops = 1 << std::max<int64_t>(0, vlmul + 1);
+    if ((VD % alignToPowerOfTwo(num_microops)) != 0) {
+        std::string error =
+            csprintf("Unaligned VD group in Widening op\n");
+        return std::make_shared<IllegalInstFault>(error, machInst);
+    }
+    if ((VS2 <= VD) && (VD < (VS2 + num_microops - 1))) {
+        // A destination vector register group can overlap a source vector
+        // register group if The destination EEW is greater than the source
+        // EEW, the source EMUL is at least 1, and the overlap is in the
+        // highest- numbered part of the destination register group.
+        std::string error =
+            csprintf("Unsupported overlap in VS2 and VD2 for Widening op\n");
+        return std::make_shared<IllegalInstFault>(error, machInst);
+    }
+
+    const int32_t t_micro_vlmax = vtype_VLMAX(machInst.vtype8, true);
+    const int32_t micro_vlmax = vlmul < 0 ? t_micro_vlmax : t_micro_vlmax / 2;
+    [[maybe_unused]] const size_t offset =
+        (this->microIdx % 2 == 0) ? 0 : micro_vlmax;
+
+    %(op_decl)s;
+    %(op_rd)s;
+    %(vm_decl_rd)s;
+    %(copy_old_vd)s;
+    %(code)s;
+    %(op_wb)s;
+    return NoFault;
+}
+
+}};
+
+def template VectorIntNarrowingMicroExecute {{
+
+template <typename ElemType>
+Fault
+%(class_name)s<ElemType>::execute(ExecContext* xc,
+                                  Trace::InstRecord* traceData) const
+{
+    using vu [[maybe_unused]] = std::make_unsigned_t<ElemType>;
+    using vi [[maybe_unused]] = std::make_signed_t<ElemType>;
+    using vwu [[maybe_unused]] = typename double_width<vu>::type;
+    using vwi [[maybe_unused]] = typename double_width<vi>::type;
+    [[maybe_unused]] constexpr size_t sew = sizeof(vu) * 8;
+
+    if (machInst.vill)
+        return std::make_shared<IllegalInstFault>("VILL is set", machInst);
+
+    const int64_t vlmul = vtype_vlmul(machInst.vtype8);
+    const uint32_t num_microops = 1 << std::max<int64_t>(0, vlmul + 1);
+    if ((VS2 % alignToPowerOfTwo(num_microops)) != 0) {
+        std::string error =
+            csprintf("Unaligned VS2 group in Narrowing op\n");
+        return std::make_shared<IllegalInstFault>(error, machInst);
+    }
+    if ((VS2 < VD) && (VD <= (VS2 + num_microops - 1))) {
```

Comment:
I see several duplicate codes, maybe we can use macros to avoid it? Also, we should use machInst.vs2 for the sake of consistency.

Reply:
I've replaced the code with macros.